### PR TITLE
Ellipse

### LINKED ISSUE #283: Add ellipse primitive
Similar to `Circle`, defined by top left as per #274, but with a width/height given as a `Size`. E.g.:

```rust
// Create an ellipse 10px wide, 20px tall
let ellipse = Ellipse::new(Point::zero(), Size::new(10, 20));

// Create a circle 20px in diameter
let ellipse = Ellipse::new(Point::zero(), Size::new(20, 20));
```

Questions:

1. [x] ~Should `Circle` become a simple wrapping struct to `Ellipse`, or should it be removed entirely? IMO `Circle` should be kept in.~ `Circle` will become a thin wrapper over `Ellipse`.

### PR DESCRIPTION
**Pull request by jamwaffles**
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a simulator example(s) where applicable
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

Adds support for the `Ellipse` primitive.

Closes #283

### CODE REVIEW

**Comment by jamwaffles** [comment]
@rfuest I think this is ready for review now. Some notes:

1. I didn't replace `Circle`s implementation with `Ellipse` as `Ellipse` is about 200% slower. It would be good if you could confirm this by running `cargo bench circle`, then changing the bench to the following code and running it again:

    ```diff
    fn filled_circle(c: &mut Criterion) {
        c.bench_function("filled circle", |b| {
            let style = PrimitiveStyleBuilder::new()
                .fill_color(Gray8::new(1))
                .stroke_color(Gray8::new(10))
                .stroke_width(1)
                .build();

    -       let object = &Circle::new(Point::new(100, 100), 100).into_styled(style);
    +       let object = &Ellipse::new(Point::new(100, 100), Size::new(100, 100)).into_styled(style);

            b.iter(|| object.into_iter().collect::<Vec<Pixel<Gray8>>>())
        });
    }
    ```

    Optimisations accepted ;)

2. I think this is squashable and can be cherry-picked into the `v0.6` branch without splitting anything out, but I'm not sure what's been added recently that's breaking. If something is in this PR that's breaking for 0.6 I'll split it into two or more commits and we can rebase this change instead of squashing.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+    pub const fn new(top_left: Point, size: Size) -> Self {
+        Ellipse { top_left, size }
+    }
+
+    /// Create a new ellipse centered around a given point with a specific size
+    pub fn with_center(center: Point, size: Size) -> Self {
+        let offset = Size::new(
+            size.width.saturating_sub(1) / 2,
+            size.height.saturating_sub(1) / 2,
+        );
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:74]:
```suggestion
        let offset = size.saturating_sub(Size::new(1, 1)) / 2;
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    pub fn center(&self) -> Point {
+        self.bounding_box().center()
+    }
+
+    /// Return the center point of the ellipse scaled by a factor of 2
+    ///
+    /// This method is used to accurately calculate the outside edge of the ellipse.
+    /// The result is not equivalent to `self.center() * 2` because of rounding.
+    fn center_2x(&self) -> Point {
+        let radius = self.size - Size::new(1, 1);
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:90]:
```suggestion
        let radius = self.size.saturating_sub(Size::new(1, 1));
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        let has_fill_color = styled.style.fill_color.is_some();
+        let has_stroke_color = styled.style.stroke_color.is_some();
+
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = if has_stroke_color {
+            styled.style.stroke_width
+        } else {
+            0
+        };
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:219]:
```suggestion
        let stroke_width = styled.style.effective_stroke_width();
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+
+        let outer_size = styled.primitive.size;
+        let inner_size = Size::new(
+            outer_size.width.saturating_sub(2 * stroke_width),
+            outer_size.height.saturating_sub(2 * stroke_width),
+        );
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:225]:
```suggestion
        let inner_size = outer_size.saturating_sub_(Size::new(2 * stroke_width, 2 * stroke_width));
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+        let iter = if has_stroke_color || has_fill_color {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:226]:
```suggestion
        let iter = if !style.is_transparent() {
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+        let threshold = if diameter <= 4 {
+            diameter.pow(2) - diameter
+        } else {
+            diameter.pow(2)
+        } as i32;
+
+        (x + y) - threshold
+    } else {
+        let a = a.pow(2) as i32;
+        let b = b.pow(2) as i32;
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:308]: These don't change and could be stored in the iter instead of `size` to improve performance.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (negative), on (0) or outside (positive) an ellipse centered around
+/// `(0, 0)` with width and height defined by the `size` parameter.
+fn size_to_threshold(size: &Size, point: &Point) -> i32 {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:284]: Is there a reason to use references here? Most other code in e-g doesn't use references for `Size` and `Point`.

**Comment by rfuest** [comment]
> 1. I didn't replace `Circle`s implementation with `Ellipse` as `Ellipse` is about 200% slower. It would be good if you could confirm this by running `cargo bench circle`, then changing the bench to the following code and running it again:

I get a change of +157% in runtime.

**Comment by rfuest** [comment]
Did you consider this https://github.com/jamwaffles/embedded-graphics/issues/283#issuecomment-610284009?

**Comment by jamwaffles** [comment]
Thanks for the quick review! I'll address those comments shortly.

> Did you consider this https://github.com/jamwaffles/embedded-graphics/issues/283#issuecomment-610284009?

I did, and I'd like to leave that for another day. We can't rotate rectangles either and I'd like to find a general solution.

**Comment by jamwaffles** [comment]
> I get a change of +157% in runtime.

Ouch

**Comment by rfuest** [comment]
> I did, and I'd like to leave that for another day. We can't rotate rectangles either and I'd like to find a general solution.

OK, good point. Can you open an issue for that?

**Comment by jamwaffles** [comment]
Sure, I've opened #321 with some spitball comments

File: embedded-graphics/src/primitives/ellipse.rs (hunk 6 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs:308]: This would require storing and passing in the non-squared version of `a` as that's used in the optimised circle path. I tried adding that in but it's not very clean.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 7 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: Thought it might improve performance but benchmarking this compared to passing by copy didn't change much. I've removed the references.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 6 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: Ok, I've pushed a change that adds a `diameter` argument to the function. A slight performance improvement. Let me know what you think.

**Comment by jamwaffles** [comment]
Changes addressed, ready for another look :+1:

**Comment by rfuest** [comment]
Instead of copying the tests from `circle.rs` it would be easier to simply compare the result of drawing a circle and an ellipse: https://gist.github.com/rfuest/2c6918cce6dc9c8a1ac92c5ab28968c0

This makes it easier to test different styles and that revealed that there seems to be a difference for the stroked styles.

Could you also add tests for the `Points` iterator and the `ContainsPoint` impl?

**Comment by jamwaffles** [comment]
> it would be easier to simply compare the result of drawing a circle and an ellipse

Very neat, thanks for the suggestion. I've copy-pasta'd your gist into the PR along with adding the other tests as requested.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+        } else {
+            diameter.pow(2)
+        } as i32;
+
+        (x + y) < threshold
+    } else {
+        let a = a as i32;
+        let b = b as i32;
+
+        b * x + a * y - b * a <= 0
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:299]: I think this should also be `< 0`.
If you set a wide border in the demo app and reduce the width or hight to `< 2 * stroke_width` you'll see a glitch appear if this is `<= 0`.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    let x = x.pow(2);
+    let y = y.pow(2);
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        let threshold = if diameter <= 4 {
+            diameter.pow(2) - diameter / 2
+        } else {
+            diameter.pow(2)
+        } as i32;
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:292]: You can refactor this code by computing this in the constructor and pass `threshold` instead of `diameter`. For the non circle case you can rearrange the formula to ` b * x + a * y <  threshold` and pass `a * b` as `threshold`.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (`true`) or outside (`false`) an ellipse centered around `(0, 0)` with
+/// width and height defined by the `size` parameter.
+fn size_to_threshold(size: Size, point: Point, diameter: u32) -> bool {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:275]: Now that the function returns a `bool` the function name doesn't seem right anymore.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+        })
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:268]: To be consistent we should probably add `DrawTarget::draw_ellipse`. Do you want to do this in this PR, later or not at all?

File: embedded-graphics/src/primitives/ellipse.rs [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self {
+            inner_size,
+            inner_color,
+            outer_color,
+            center,
+            inner_width,
+            ..
+        } = *self;
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:247]: I think this will copy more values than necessary per `next` call, which could impact performance. Using `for point in self.iter.by_ref() { ... }` isn't as nice as using `find_map` but you can use `self` inside the for loop.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+    }
+}
+
+impl Iterator for Points {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ellipse = self.ellipse;
+
+        self.iter.find(|point| ellipse.contains(*point))
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:182]: The performance of this could be improved by storing `center_2x` and the squared size in the `Points` struct and using `size_to_threshold`, like in the `StyledEllipseIterator`.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+}
+
+/// Pixel iterator for each pixel in the ellipse border
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    iter: Points,
+    outer_size: Size,
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:193]: This field is never used.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 11 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: I've added it but in a separate commit with it's own changelog item. I haven't seen a display that can natively draw an ellipse, but it's here if people want it. I'll rebase this branch into master when it's done to preserve both commits.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 8 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: Reproduced and fixed, thanks for catching that. I've added another test to catch any regressions.

**Comment by jamwaffles** [comment]
Those improvements look good. I've integrated them into the PR.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_ellipse(self)
+    }
+}
+
+fn compute_threshold(size: Size, diameter: u32) -> u32 {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:281]: The function could be changed to `fn compute_threshold(size: Size) -> (Size, u32)` by doing the squaring of size in this function. This would remove the need to pass a squared and non squared version of diameter and reduce some code duplication.

A slightly fancier version of this would be to use a struct instead of `(Size, u32)` which could also be used to implement the `is_point_inside_ellipse` function.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+        height: b,
+    } = size;
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        if diameter <= 4 {
+            diameter.pow(2) - diameter / 2
+        } else {
+            diameter.pow(2)
+        }
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:293]: We could share this code with `circle.rs` by making `diameter_to_threshold` in `circle.rs` `pub(crate)`. I'm not sure if that's an huge improvement, but it would make sure that they stay equal.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+    } = size;
+
+    let Point { x, y } = point;
+
+    let x = x.pow(2) as u32;
+    let y = y.pow(2) as u32;
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        (x + y) < threshold
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:315]:
```suggestion
        x + y < threshold
```

File: embedded-graphics/src/primitives/ellipse.rs [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+        Ellipse::new(Point::new(0, 0), size)
+            .into_styled(style)
+            .draw(&mut display)
+            .unwrap();
+
+        assert_eq!(display, MockDisplay::from_pattern(pattern));
+    }
+
+    fn test_circles(style: PrimitiveStyle<BinaryColor>) {
+        for diameter in 1..50 {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:353]: Why did you change this to start from `1` instead of `0`? IMO diameter `0` should also match.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+    type Item = Pixel<C>;
+    type IntoIter = StyledEllipseIterator<C>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        StyledEllipseIterator::new(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:334]: If I haven't missed something, all tests seem to use `top_left = Point::zero()`. IMO at least one should use a different value for `top_left`.
This could maybe be combined with a test for `translate`, which isn't tested yet.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 16 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: Good idea. I went with `pub(in crate::primitives)` in `circle.rs` to keep it more tightly scoped.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 15 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs]: Good suggestion, changed. Adding a struct sure is fancy but this is a pretty small internal method so not worth the effort IMO.

File: embedded-graphics/src/primitives/ellipse.rs (hunk 18 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs:348]: Oops, that was a debugging artifact left over. It should indeed by `0..50`. Changed!

File: embedded-graphics/src/primitives/ellipse.rs (hunk 19 above)
Comment (jamwaffles) [reply embedded-graphics/src/primitives/ellipse.rs:329]: I've added a combined non-zero origin and translate test.

File: embedded-graphics/src/primitives/ellipse.rs [hunk 20]
```diff
@@ -0,0 +1,543 @@
...
+    type PointsIter = Points;
+
+    fn points(&self) -> Self::PointsIter {
+        Points::new(self)
+    }
+}
+
+impl ContainsPoint for Ellipse {
+    fn contains(&self, point: Point) -> bool {
+        // let size_sq = Size::new(self.size.width.pow(2), self.size.height.pow(2));
```
Comment (rfuest) [inline embedded-graphics/src/primitives/ellipse.rs:103]:
```suggestion
```

File: embedded-graphics/src/primitives/ellipse.rs (hunk 18 above)
Comment (rfuest) [reply embedded-graphics/src/primitives/ellipse.rs:348]: You only changed this in `circles_points` and not in `test_circles`.

**Comment by jamwaffles** [comment]
ralf

**Comment by jamwaffles** [comment]
:heart: :pray:
